content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md

```diff
@@ -0,0 +1,742 @@
+          path: /etc/kubernetes/pki
+          type: DirectoryOrCreate
+        name: k8s-certs
+      - hostPath:
+          path: /etc/ssl/certs
+          type: DirectoryOrCreate
+        name: ca-certs
+      - name: cloud-config-volume
+        secret:
+          secretName: cloud-config
+      - name: ca-cert
+        secret:
+          secretName: openstack-ca-cert
+```
+
+When the controller manager is running, it will query OpenStack to get information about the nodes and remove the taint. In the node info you'll see the VM's UUID in OpenStack.
+```
+# kubectl describe no master1
+Name:               master1
+Roles:              master
+......
+Taints:             node-role.kubernetes.io/master:NoSchedule
+                    node.kubernetes.io/not-ready:NoSchedule
+......
+sage:docker: network plugin is not ready: cni config uninitialized
+......
+PodCIDR:                     10.224.0.0/24
+ProviderID:                  openstack:///548e3c46-2477-4ce2-968b-3de1314560a5
+
+```
+Now install your favourite CNI and the master node will become ready.
+
+For example, to install weave net, run this command
+```
+kubectl apply -f "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')"
+```
+
+Next we'll set up worker nodes.
+
+Firstly install docker and kubeadm in the same way as how they were installed in the master. 
+To join them to the cluster we need a token and ca cert hash from the output of master installation. 
+If it is expired or lost we can recreate it using these commands.
+
+```
+# check if token is expired
+kubeadm token list
+
+# re-create token and show join command
+kubeadm token create --print-join-command
+
+```
+
+Create kubeadm-config.yml for worker nodes with the above token and ca cert hash.
+```
+apiVersion: kubeadm.k8s.io/v1beta2
+discovery:
+  bootstrapToken:
+    apiServerEndpoint: 192.168.1.7:6443
+    token: 0c0z4p.dnafh6vnmouus569
+    caCertHashes: ["sha256:fcb3e956a6880c05fc9d09714424b827f57a6fdc8afc44497180905946527adf"]
+kind: JoinConfiguration
+nodeRegistration:
+  kubeletExtraArgs:
+    cloud-provider: "external"
+
+```
+apiServerEndpoint is the master node, token and caCertHashes can be taken from the join command printed in the output of 'kubeadm token create' command.
+
+Run kubeadm and the worker nodes will be joined to the cluster.
+```
+kubeadm join  --config kubeadm-config.yml 
+```
+
+At this stage we'll have a working Kubernetes cluster with an external OpenStack cloud provider.
+The provider tells Kubernetes about the mapping between Kubernetes nodes and OpenStack VMs.
+If Kubernetes wants to attach a persistent volume to a pod, it can find out which OpenStack VM the pod is running on from the mapping, and attach the underlying OpenStack volume to the VM accordingly.
+
+The integration with Cinder is provided by an external Cinder CSI plugin, as described in https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-cinder-csi-plugin.md
```

Comment:
```suggestion
The integration with Cinder is provided by an external Cinder CSI plugin, as described in [the Cinder CSI documentation](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-cinder-csi-plugin.md)
```

It's generally better to put the link behind some text instead of providing a raw link.